### ISSUE TITLE
Search for a space from 'My Spaces' (8)

### ISSUE THREAD

**Issue by catrobson** [openshiftio/openshift.io#3655]
### Description
Currently the only way for users to search for a space is in the "All Spaces" page.  This page is old and meant to be deprecated.  Before we can remove this page, we need to relocate the functionality of searching for a space, requesting an invite, etc.  This story will be determining how we can change the current 'my spaces' page to include a search.

### Parent Issue(s)
- [past link here]() or None

### Verification Criteria
- [x] Update My Spaces designs to include a space search capability.  Keep in Google Drive.  Upload changes to InVision.
- [x] Review with UXD team
- [x] Review with Auth, Platform teams in meetings where possible.  Also share with devtool-saas mailing list.

**Comment by jiekang** [comment]
Likely to be relevant? https://github.com/openshiftio/openshift.io/issues/477

**Comment by sdash-redhat** [comment]
@catrobson Is there a link to the latest design files for the space that the current implementation is based on?

**Comment by sdash-redhat** [comment]
@catrobson I found one active link in #477 i.e. https://redhat.invisionapp.com/share/DQDSL22JA#/screens/258353405
Is this the latest one?

@catrobson @muruGanesan @joshuawilson  @bmicklea 
Also, to understand some background on the earlier discussions that might have touched upon the search aspect  --

what was stakeholder's opinion on the using a global search bar in the main navigation bar that allows to search based on context? Are we looking into that aspect in this story?

Or alternatively ...are we expecting a separate search control for explicitly searching the spaces in the UI?

**Comment by joshuawilson** [comment]
I think that if we had a global search that would cover the need to search here. However, I think we can make this minor change much faster and should do this first. Then we can remove it when the global is in place.

**Comment by sdash-redhat** [comment]
@aslakknutsen Your thoughts will be valuable in this context. Do you believe, for a long-term perspective, we should explore the global serach option (maybe as a separate story)... Also any other thoughts in this regard?

**Comment by sdash-redhat** [comment]
Based on my discussions with @catrobson, @joshuawilson  and @aslakknutsen  it appears that it makes more sense to have a global search that would be more meaningful for the users in various contexts of what they are looking for. That should be somewhat a long-term and recommended approach, and it is the right time we can explore the scope and behaviour and design of global search. 

Meanwhile, for the current story, we will be looking into the short-term solution to bring the search form "All Spaces" to "My Spaces" to retain the existing functionality. Later this can be replaced when a global search is available. 



++ @serenamarie125 @muruGanesan

**Comment by sdash-redhat** [comment]
Here is the Invision link of the first draft of the interaction with 2 options for short-term. https://redhat.invisionapp.com/share/E4LH2OCDC7A
Please feel free to go through and share your thoughts.

@catrobson @serenamarie125 @joshuawilson @aslakknutsen @lwrigh @mceledonia  @bdellasc @dlabrecq  @muruGanesan @sunilmalagi @Essjaysee

**Comment by sdash-redhat** [comment]
Here is the updated version (0.2)  : [https://redhat.invisionapp.com/share/P7LNZNZUZ9Y](https://redhat.invisionapp.com/share/P7LNZNZUZ9Y)

**Comment by sdash-redhat** [comment]
UXD Review meeting notes (from today): 
1. Type-ahead shown in the experience is to be marked as a stretch goal for implementation. 
2. The Grey Background in Modal -- should it be white? -- While implementing in HTML, the team can play around with the background in modal.

**Comment by andrewazores** [comment]
I'm prototyping implementation for this to ensure current feasibility from the engineering perspective. The querying capability for spaces is, as far as I know, restricted to querying by space name. It also doesn't always behave exactly as I would expect - for example, querying for "test-space" returns one set of results, but querying for just "test" returns a *smaller* set of results, which does not contain any of the results in the first set.

Is there an issue open for the task or story of implementing this yet, or is it still only tracked here for the design work?

**Comment by sdash-redhat** [comment]
@catrobson  Looks like we have an agreed UX/design as of now. So closing this issue. 
If we get further feedback/ we can reopen or use a different story. 

The technical prototyping /feasibility aspect can be handled in a different story as suggested by @andrewazores

**Comment by andrewazores** [comment]
Has a story for the technical side been opened anywhere? I have some details regarding the current querying capability, and some questions about the typeahead behaviour in the design as well.

**Comment by sdash-redhat** [comment]
@andrewazores I'm reopening the issue. 
Please share your thoughts on the design and related aspects here. 
Also, you can provide feedback on the design in the InVision document.

**Comment by rohitkrai03** [comment]
@andrewazores Have you decided on implementation details for the typeahead search box? After going through the designs I think `ng-select` component can be reused for that which is currently being used in add-collaborators component - https://github.com/fabric8-ui/fabric8-ui/blob/2f3979be26235950053faed7325947f5cc3d3427/src/app/space/settings/collaborators/add-collaborators-dialog/add-collaborators-dialog.component.html#L11

**Comment by andrewazores** [comment]
My concern with the typeahead is that the result set can be very large (eg. searching for "test" currently returns 1900+ results), and so the results are paginated. This means that the typeahead results will probably be restricted to the first page of results for the current search term. I don't know what kind of filtering or sorting is supported on the backend, so this might feel wonky if the filtering and sorting isn't done right. The first page of results for "test" are all going to look very similar since it's such a generic name. When an actual search is performed and the list view is populated this is not a problem, because the non-unique "test" name becomes unique when displayed alongside the creator's username/avatar, and the description may help as well. In the search typeahead, there could very well be 20 non-distinct "test" space names comprising the entire page of results.

Additionally, how long should the delay be between the user finishing typing and the appearing of the typeahead be? There needs to be a buffered delay time so that we aren't firing off a query to the backend on every single keystroke a user enters into the search box.

As for the querying API itself - what kind of search functionality is expected here? The current API appears to return results where the space's name or description match the given search term, and this doesn't seem to be configurable.

**Comment by andrewazores** [comment]
Right, that's what I figured would be the bare minimum. But then we have a case where when you type in "tes", the typeahead will only display a single suggestion based on the first page of results: "test". It wouldn't display "testA" because that wouldn't be found until the second, or third, or nth page of results, which won't be retrieved until the user has entered a search term that matches past "test" - ie, until they actually write "testA" themselves.

The example of "test" is specifically one that currently has and probably always will have a huge number of results, so this is a sort of extreme case to examine the functionality under, but that's the purpose of the exercise.

**Comment by rohitkrai03** [comment]
@andrewazores @sdash-redhat Instead of showing just one result for 20 non-distinct ones we can probably add username and user avatar in the result list of typeahead. That would solve the problem @andrewazores mentioned in the above comment. 

Also, for a very large set of result we can have an infinite scroll functionality implemented.
For the delay we can simply add a debounce of 200ms for input event which would be sufficient restrict multiple calls to backend.

What do you guys think?

**Comment by sdash-redhat** [comment]
@andrewazores 
Good points, just thinking aloud, see if it helps -- 

I think now we are entering into how the search algorithm is getting the data from the backend to form ... 

"...that wouldn't be found until the second, or third, or nth page of results, which won't be retrieved until the user has entered a search term that matches past "test" - ie, until they actually write "testA" themselves..."


Maybe the algorithm should keep some flag to count if 20 unique matches are found (or end of record has reached) before sending that list to display in typeahead...

On second thought ...maybe there should be a high level cached/pre-generated list of space names available and when user types in it only get string match from this list only in the front end... that might solve the issues like 

1. almost all major/frequently searched/popular (depends on how the algorithm is defined)  will be covered in the list based match ...if something that does not match that may anyway would be searched eventually by the user ...

2. This will limit the call to backend multiple times, ...for instance, let's say the list is pre-fetched, there will not be any further call needed to backend just to populate type ahead... Variations of this loading and updating of this list can be played around to find the suitable approach to fit our need. 

@rohitkrai03 
Definitely a different angle..., that would depend on the search algorithm you guys choose I believe -- for instance, do we consider only show typeahead for space names or any other additional parameters like username etc. etc.  I feel that might be a little more overhead for just type ahead... but a purely technical call I believe from a feasibility perspective.

**Comment by andrewazores** [comment]
Okay. Lots of food for thought here. I think the typeahead functionality will take some extra time and effort to implement because it requires some expanded backend support specific to that functionality.

Is the intent of the results list to be an infinite scroll, rather than Patternfly pagination? I have it prototyped with a pagination component at the moment - see the gif below. It's still missing backend sorting of results.

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/3787464/42643556-6b0d001e-85c7-11e8-8d52-22fdccb3bf5c.gif)

**Comment by sdash-redhat** [comment]
Awesome Andrew! Looks great.  
Yes, the  initial intent was to have an infinite scroll, as users have better experiences with this. 

P.S. I feel if we can increase the modal width , may that will help show longer description that might be there in the result items under the spce names.

**Comment by andrewazores** [comment]
Is the increased width idea to help maintain a constant modal height when longer descriptions are shown? There are a few spaces I've seen with really, really long descriptions where no reasonable modal width would end up helping. Maybe there's another discussion to be had around a trimming or expansion behaviour of the space descriptions here.

I'll continue looking into the backend API for sorting and filtering. As-is it won't play so nicely with infinite scroll, since the results aren't pre-sorted. If we just do the sorting on the frontend then we will end up with situations where scrolling down will insert new results into the list above where the user is currently looking.

**Comment by andrewazores** [comment]
And PS, what are your thoughts on the button used to summon the modal? I took a little liberty with that and used a button that mimics the style of the "Create Space" one.

**Comment by sdash-redhat** [comment]
@andrewazores True. We need to maintain a constant height at all time irrespective of the number of results displayed. Wider width will help in providing descent whitespace among the each item in a line. It would help us maintain a decent 

https://redhat.invisionapp.com/share/23CC7R3QZ#/screens/235814943_Slide01

Regarding the visual aspect , size dimension etc, Sunil can help ....

@sunilmalagi Please refer the thread and help through suggesting 
1. the button size and look and feel 
2. The specs of the modal ..how much wider it can be ...

**Comment by andrewazores** [comment]
Here's a quick update for where I'm at right now. Still needs visual tweaks, needs sorting support (backend and frontend), and we still need to settle how typeahead will work in practical terms. I think typeahead will end up being a future enhancement.

![ezgif-2-acd632a934](https://user-images.githubusercontent.com/3787464/43014744-30618d5e-8c1b-11e8-99ea-35b5eee804ad.gif)

**Comment by sdash-redhat** [comment]
@andrewazores Looks really great!!! 

@sunilmalagi is creating specs to help you tweak the minor visual glitches like spacing and padding. He will be sharing with you shortly today. Hope that will help :)

Thanks again for the great stuff!

**Comment by sunilmalagi** [comment]
@sdash-redhat @andrewazores 
Here is the visual screen for Space list with specification (total 2 screens), please check the below link and let me know if any further details required:
https://redhat.invisionapp.com/share/2HN3WIHAGQ6#/310210007_SpaceList_01

**Comment by andrewazores** [comment]
Here's another update. I need to do some refactoring and write tests, and I'd like someone else who's better at CSS help me clean that up as well.

Still no support for sorting or typeahead, still thinking those should be done as future followups.

![ezgif-5-563e258c3b](https://user-images.githubusercontent.com/3787464/43343519-836f4d9e-91b4-11e8-9822-7e18ea0acfba.gif)

**Comment by andrewazores** [comment]
This feature as it currently stands in production (see PR linked above or GIF above) was just demoed in sprint review, and a good point was raised: this new view provides no way for owners to delete their Spaces, but this is functionality is present in the View All Spaces view that this is intended to replace. What is the intended path going forward for owners to delete their Spaces? Should this be included somehow in the My Spaces Search dialog?

**Comment by sdash-redhat** [comment]
@andrewazores That's a good point! I believe that should the case, the user should be able to delete in My Space.  
@catrobson @serenamarie125 @muruGanesan what's your thought on this?

**Comment by catrobson** [comment]
@sdash-redhat You should only be able to delete spaces that you own, so I don't think we need this in the search dialog.  We should only need to allow for deletion on the my spaces list.  I did notice that we do have "remove space" in the kabob menu there now.  In the design we actually specified two behaviors:  "Delete space" (when it is owned by me), and "Leave space" (When it has been added to my spaces that I'm a collaborator on, but I no longer to be part of that space I can leave the space from which will remove it from the list, but I won't be deleting the space itself).  I think we need to correct this implementation, and that should cover our use cases.  Design captured here:   https://redhat.invisionapp.com/share/DQDSL22JA#/screens/261389228

**Comment by serenamarie125** [comment]
@sdash-redhat please put this on the agenda for the Dec 19 stakeholder review, so we can be sure that it has been implemented already

**Comment by sdash-redhat** [comment]
@serenamarie125 This was later addressed by  **Improve the My Spaces page #4326** which was reviewed and closed.

**Comment by sdash-redhat** [comment]
Based on today's stakeholder meeting: 
1. The new IA should be implemented. 
2. The upfront available "Create Application" must move to kebab menu 
3. Need to keep an eye for mobile/small screen touch-screen aspect. 

@serenamarie125 : please provide the new IA template to use. 

Moving the status to in-progress.

**Comment by sdash-redhat** [comment]
Based on the discussion made the changes. The updated workflow is in the following: 

**https://redhat.invisionapp.com/share/VYPRH2I793F**

Please review and let me know your thoughts. 
@serenamarie125 @catrobson

**Comment by sdash-redhat** [comment]
Based on the feedback need to change the following 2 and add it to stakeholder's meeting: 

1. Change the title of the modal to "Search Spaces" - that one was issued out
2. Use Parvathyvr's reference screen as the parent page. 

@serenamarie125 @parvathyvr

**Comment by parvathyvr** [comment]
https://redhat.invisionapp.com/share/Y7OC1HORWQC#/screens/322844508 
screen for reference-done as part of #4326

**Comment by joshuawilson** [comment]
At this point, this feels like a bug. If UX is done, please assign to UI team.

**Comment by sdash-redhat** [comment]
Updated the invasion on the same link. 
**[https://redhat.invisionapp.com/share/VYPRH2I793F](https://redhat.invisionapp.com/share/VYPRH2I793F)**

**Comment by joshuawilson** [comment]
Looks fine to me.

**Comment by sdash-redhat** [comment]
Thanks, Joshua :)
